Issue1658 prevent nodes signing multiple blocks at the same height

hewison-chris opened the pull request:
This creates noise `r` for block signing for each node by adding `rc` to `preimage` reduced to Scalar.

linked0 reviewed (approved):
LGTM

hewison-chris commented:
I changed the issue requirements and also this PR to reflect the changes. It should use block hash for payload but random seed for  noise `r`.

linked0 reviewed (approved):
LGTM

AndrejMitrovic commented:
`Fixes #1658` in the commit message would help to link the issue.

But where is the test-case that signing an alternate block reveals the private key?

hewison-chris commented:
> `Fixes #1658` in the commit message would help to link the issue.
> 
> But where is the test-case that signing an alternate block reveals the private key?

There is https://github.com/bpfkorea/agora/issues/408 which would include that. However I could still add a test in this PR.

AndrejMitrovic commented:
Also I'm not sure this actually solves the problem. Because it uses the collective sum of the preimages rather than the node's own preimage only. For example, If you have 5 nodes:

[A, B, C, D, E]

And nodes [A, B, C] sign one block at height 1:

A: preimage 1
B: preimage 2
C: preimage 3

random_seed = H(1, 2, 3)

But nodes [C, D, E] decide to sign an alternate block at height 1:

C: preimage 3
D: preimage 4
E: preimage 5

random_seed = H(3, 4, 5)

The random seed will be different for the same block height. Therefore the private key will not be revealed? Am I off base here?

AndrejMitrovic commented:
It would help resolve all doubts if we had the test-cases demonstrating the behavior.

hewison-chris commented:
> It would help resolve all doubts if we had the test-cases demonstrating the behavior.

I think you are right in that my assumption of unique random seed is flawed. I will add a test and try a different approach. Perhaps adding the `height.hashFull()` to the commitment to generate a unique `r` for each block height is a valid approach.

hewison-chris commented on source/agora/consensus/data/Block.d:
```diff
@@ -824,3 +824,53 @@ version (unittest)
+    assert(c1 == c + c2);
+
+    Scalar secret = s * c.invert();
+    assert(secret == v,
```
I can not figure out why this is not working.

AndrejMitrovic replied:
I'll take a deeper look at this tomorrow.

AndrejMitrovic replied:
Where did you get this formula from?

hewison-chris replied:
https://tlu.tarilabs.com/cryptography/digital_signatures/introduction_schnorr_signatures.html#replay-attacks

hewison-chris replied:
Also https://en.wikipedia.org/wiki/Schnorr_signature

AndrejMitrovic commented on source/agora/consensus/data/Block.d:
```diff
@@ -824,3 +824,53 @@ version (unittest)
+    auto block2 = GenesisBlock.makeNewBlock(
+        genesisSpendable().take(3).map!(txb => txb.sign()).array(), 10, random_seed2);
+
+    Scalar v = genesis_validator_keys[0].secret;
```
This is the problematic part. This secret is not on the curve at all. Probably something to do with the recent crypto changes. If you change it to `Scalar v = Scalar.random();` it will work fine.

AndrejMitrovic replied:
Full example where it works:

```D
import std.stdio;

Scalar v = Scalar.random();
//Scalar v = genesis_validator_keys[0].secret;

Scalar rc = Scalar(hashMulti(v, "consensus.signature.noise", 0));
Scalar r = rc + Scalar(block1.header.height.hashFull());

Scalar c1 = Scalar(hashFull(block1));
Scalar c2 = Scalar(hashFull(block2));

// Sign with same r twice
Scalar s1 = r + (v * c1);
Scalar s2 = r + (v * c2);

Scalar sig = s1 - s2;

writefln("Is it: %s", v == (s1 - s2) * (c1 - c2).invert);
assert(v == (s1 - s2) * (c1 - c2).invert);
```

hewison-chris replied:
> Full example where it works:
> 
> ```d
> import std.stdio;
> 
> Scalar v = Scalar.random();
> //Scalar v = genesis_validator_keys[0].secret;
> 
> Scalar rc = Scalar(hashMulti(v, "consensus.signature.noise", 0));
> Scalar r = rc + Scalar(block1.header.height.hashFull());
> 
> Scalar c1 = Scalar(hashFull(block1));
> Scalar c2 = Scalar(hashFull(block2));
> 
> // Sign with same r twice
> Scalar s1 = r + (v * c1);
> Scalar s2 = r + (v * c2);
> 
> Scalar sig = s1 - s2;
> 
> writefln("Is it: %s", v == (s1 - s2) * (c1 - c2).invert);
> assert(v == (s1 - s2) * (c1 - c2).invert);
> ```

Thanks that is the problem. if I add a check I see the secret key is not a valid Scalar (which it should be):-
```
Scalar v = genesis_validator_keys[0].secret;
assert(v.isValid(), "v is not a valid Scalar!");
```

hewison-chris replied:
The problem is that the `WellKnownKeys` have SecretKey values which are not valid Scalar values. I will create a PR to address that specifically first.

Geod24 commented on source/agora/consensus/data/Block.d:
```diff
@@ -796,7 +796,7 @@ version (unittest)
             const Scalar rc = Scalar(hashMulti(key.secret, "consensus.signature.noise",
                 cycleForValidator(key.address)));
-            const Scalar r = rc + challenge; // make it unique each challenge
+            const Scalar r = rc + Scalar(block.header.height.hashFull());
```
Why do we hash the height ?? It should be the pre-image.

hewison-chris replied:
Using the height hashed is a quick win solution (from using the block hash) to prevent nodes signing more than one block at a certain height. Can you please explain why using the `pre-image` is preferred here?

hewison-chris replied:
After discussion we decided to use pre-image to be more secure as an attack might still be possible as the hashed height is predictable for all blocks whereas the pre-image are only revealed for several blocks ahead at most.

hewison-chris committed:
Enable safe subtraction from Height

hewison-chris commented:
Still need to fix up those difficult block validation unit tests that need pre-images now.

hewison-chris commented on source/agora/consensus/data/Block.d:
```diff
@@ -818,3 +818,53 @@ version (unittest)
+    assert(v.isValid(), "v is not a valid Scalar!");
+    Point V = v.toPoint();
+    const Scalar rc = Scalar(hashMulti(v, "consensus.signature.noise", 0));
+    const Scalar r = rc + Scalar(block1.header.height.hashFull());  // TODO: Use pre-image
```
TODO still

hewison-chris committed:
Trivial: rename variable keys to key_pairs

hewison-chris committed:
Test code: compare a subset of blocks

When checking that blocks are consistent for nodes it is more
convenient to only compare a subset of the blocks.
The Simple network test has been updated to take advantage of this
and compare blocks at important heights related to validator cycle.

hewison-chris committed:
Use preimage to generate the signature noise each block

hewison-chris committed:
Trivial: add some logging for test failures

hewison-chris committed:
Trivial: correct function doc

hewison-chris commented:
As the branch name is mis-leading and the direction changed a few times I am closing this PR in favour of https://github.com/bosagora/agora/pull/1860.
